client/shared/src/search/parser/parser.ts

```diff
@@ -272,6 +272,122 @@ const scanToken = <T extends Term = Literal>(
+ *
+ * So, this scanner detects parentheses as patterns without needing the user to
+ * explicitly escape them. As such, there are cases where this scanner should
+ * not succeed:
+ *
+ * (foo or (bar and baz)) - a valid query with and/or expression groups in the query langugae
+ * (repo:foo bar baz)     - a valid query containing a recognized repo: field. Here parentheses are interpreted as a group, not a pattern.
+ */
+export const scanBalancedPattern: Parser<Literal> = (input, start) => {
+    let adjustedStart = start
+    let balanced = 0
+    let current = ''
+    const result: string[] = []
+
+    const nextChar = (): string => {
+        current = input[adjustedStart]
+        adjustedStart++
+        return current
+    }
+
+    if (!keepScanning(input, start)) {
+        return {
+            type: 'error',
+            expected: 'not a recognized filter or operator',
+            at: start,
+        }
+    }
+
+    while (input[adjustedStart] !== undefined) {
+        current = nextChar()
+        if (current === ' ' && balanced === 0) {
+            // Stop scanning a potential pattern when we see whitespace in a balanced state.
+            adjustedStart-- // Backtrack.
+            break
+        } else if (current === '(') {
+            if (!keepScanning(input, adjustedStart)) {
+                return {
+                    type: 'error',
+                    expected: 'not a recognized filter or operator',
+                    at: adjustedStart,
+                }
+            }
+            balanced++
+            result.push(current)
+        } else if (current === ')') {
+            balanced--
+            if (balanced < 0) {
+                // This paren is an unmatched closing paren, so we stop treating it as a potential
+                // pattern here--it might be closing a group.
+                adjustedStart-- // Backtrack.
+                balanced = 0 // Pattern is balanced up to this point
+                break
+            }
+            result.push(current)
+        } else if (current === ' ') {
+            if (!keepScanning(input, adjustedStart)) {
+                return {
+                    type: 'error',
+                    expected: 'not recognized filter or operator',
+                    at: adjustedStart,
+                }
+            }
+            result.push(current)
+        } else if (current === '\\') {
+            if (input[adjustedStart] !== undefined) {
+                current = nextChar()
+                // Accept anything anything escaped. The point is to consume escaped spaces like "\ "
+                // so that we don't recognize it as terminating a pattern.
+                result.push('\\', current)
+                continue
+            }
+            result.push(current)
+        } else {
+            result.push(current)
+        }
+    }
+
+    if (balanced !== 0) {
+        return {
+            type: 'error',
+            expected: 'not unbalanced parentheses',
```

Comment:
```suggestion
            expected: 'balanced parentheses',
```

Can we also add a test? This basically just returns an error when it's a term with a dangling paren (i.e. should fail given this function checks for `BALANCED` parens, right? (added a suggestion in the docstring above too if this is correct).

Reply:
Yep, added. For the unbalanced case, there will be more tests, since we will still allow `foo(` in general, but not in this function :-)